### PR TITLE
[FIX] website: ignore the scheme for page indexing

### PR DESCRIPTION
**Pull request by Guillaume-gdi**
When a user sets up a domain name on Odoo, we consider that he has a
configuration that makes only one site visible. To do this, the standard
solution is to have the following redirections:
- http://example.com => https://example.com
- https://www.example.com => https://example.com
- http://www.example.com => https://example.com

It happens that users enter something other than https://example.com in
the setting to define the domain names of their websites. If this is the
case, [this other commit] would cause an error:
- The page indexing bot went to https://example.com but since this was
not what was set in the settings, a no index was added so that the page
was not referenced.
- As soon as the indexing bot went to http://example.com,
https://www.example.com or http://www.example.com, it was redirected to
https://example.com.

As a result, the client ended up with a non-indexed website.
The purpose of [this other commit] was just to prevent double indexation
of websites (the https://example.odoo.com and the https://example.com).

After this commit, the pages will be indexed even if the scheme is not
the same as the one specified in the settings. The same goes for the
[www](http://www/). which is also ignored.

Note that this can have an undesirable effect if the client has a bad
configuration and has several sites exposed (https://example.com and
https://www.example.com for example). If this is the case, he will end
up with a site that is indexed twice.

[this other commit]: https://github.com/odoo/odoo/commit/3739d74afe824554b37b1b52ed32ada33692c01a

task-3110888

### CODE REVIEW

**Comment by Guillaume-gdi** [comment]
I hope that this will not be a problem :thinking: 
image

**Review by rdeodoo — CHANGES_REQUESTED** [review]

File: addons/website/models/ir_qweb.py
Comment (Guillaume-gdi) [reply addons/website/models/ir_qweb.py:60]: @rdeodoo 

> Also, since if this was an util, it should be an Odoo util and not a website util, as it's not related to website.

Yes I can put the util in base or something like that but I'm not sure to understand your point. Domain name seems very related to website. Isn't it ? :thinking: 

Ok for the rest :+1:

Comment (rdeodoo) [reply addons/website/models/ir_qweb.py]: Wanting to extract the "base"/"pure"/"netloc" domain from an URL is not related to the website.
Anyone might need it for any given URL, that's what I meant, if it was an util to do so as you did, it should probably be in base and just take an url as param, and website calls it passing the website domain.
(Typical pattern for most of our util has always been the same: we create in website, another module needs it and move it to base).

File: addons/website/tools.py
Comment (rdeodoo) [inline addons/website/tools.py:187]: What I would do, depending of how you judge the usefulness of this method is:
- _If judged really useful but also judged worth an util rather than copying it where needed_: Ok to keep it in website, it will saves us an approval from another team and it's not the most useful util we made so they can live without it.
Still, I'd make the `ẁww.` strip conditional like a method param `strip_www=False` and I would use the util here too: https://github.com/odoo/odoo/blame/f9d5d1919b5794623a2c6296135f0692f8979e08/addons/website/models/website.py#L973
https://github.com/odoo/odoo/blame/f9d5d1919b5794623a2c6296135f0692f8979e08/addons/website/controllers/main.py#L141
- _If judged not really useful_: I would make it a sub method of `should_index_url` and call it in there directly.

I doubt it's really usual to want to have the base domain with `www` strip, but I'll let you decide about that :+1: pick what you prefer (or suggest something else)

File: addons/website/tools.py
Comment (rdeodoo) [inline addons/website/tools.py:9]: direct import are on top, see https://github.com/odoo/odoo/pull/83508#discussion_r803814729

**Review by rdeodoo — CHANGES_REQUESTED** [review]

File: addons/website/tools.py
Comment (rdeodoo) [inline addons/website/tools.py:177]: ty for docstring, just missing the `strip_www` param

File: addons/website/models/website.py
Comment (rdeodoo) [inline addons/website/models/website.py:317]: Maybe it's worth something more than this (but a bit less detailed than commit msg) explaining that the website domain and the url "domain" should match to be considered indexable.
And that matching means being equal once scheme and potential `www` has been removed.
That this is to be able to still index badly configured domain (when the domain is not the real canonical url due to www/naked or http/https redirect).

Probably not worth mentioning the tradeoff: if a website is reachable both in www+naked or http+https, it might be indexed twice but it's still better than preventing completely the indexation of a website if the domain is misconfigured (not the real endpoint).

Because I am sure no one will understand this by reading the code.

**Review by rdeodoo — APPROVED** [review]
You can squash your commit and flag the PR as Ready for Review so I have a final check and r+ (or security check? not needed anymore I think)

**Review by rdeodoo — APPROVED** [review]
I wanted to merge but then saw it was targeting master, why? Did we discuss something I forgot?
This is quite a critical fix (it prevents wiping completely your website from crawlers/indexation), please fix it in the earliest supported version possible.
Approving this, just let me know when target is changed

File: addons/website/views/website_templates.xml
Comment (Guillaume-gdi) [inline addons/website/views/website_templates.xml:94]: >I wanted to merge but then saw it was targeting master, why? Did we discuss something I forgot?
This is quite a critical fix (it prevents wiping completely your website from crawlers/indexation), please fix it in the earliest supported version possible.
Approving this, just let me know when target is changed

@rdeodoo Isn't it a problem to do a change in this file in stable ? :thinking:

Comment (rdeodoo) [reply addons/website/views/website_templates.xml:94]: No, worst case some people won't have the fix without updating website app.
But it won't break anything else, what we can't do is remove attribute or class as it will break xpath.
Here, it's just about modifying business code logic inside a `t-if` condition, no one will have written a xpath being `[t-if:contains("website.domain not in request.httprequest.url_root")]`. And even if they did, it's out of what we support, they should use class (contain and not equal) or attribute.

**Comment by rdeodoo** [comment]
Indeed, but then seeing this commit you also need to do the exact same fix in the robots.txt ^^ (using the new website method rather than the `in` check).
I forgot about that part while reviewing your PR initially.

I don't know why I didn't want to introduce it in version earlier than master at the time, it was a bad decision.
But it's fine, let's target 15.2 to be consistant with the other fix.
We will consider backporting it if someone complains about it in an opw

**Comment by Guillaume-gdi** [comment]
@rdeodoo Done :heavy_check_mark:

**Review by rdeodoo — APPROVED** [review]
@robodoo r+

Thank you for the fix

**Pull request closed without merging**
